File: cpp98/src/Ice/ConnectionRequestHandler.cpp
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

#include <Ice/ConnectionRequestHandler.h>
#include <Ice/Proxy.h>
#include <Ice/Reference.h>
#include <Ice/ConnectionI.h>
#include <Ice/RouterInfo.h>
#include <Ice/OutgoingAsync.h>

using namespace std;
using namespace IceInternal;

ConnectionRequestHandler::ConnectionRequestHandler(const ReferencePtr& reference,
                                                   const Ice::ConnectionIPtr& connection,
                                                   bool compress) :
    RequestHandler(reference),
    _connection(connection),
    _compress(compress)
{
}

RequestHandlerPtr
ConnectionRequestHandler::update(const RequestHandlerPtr& previousHandler, const RequestHandlerPtr& newHandler)
{
    assert(previousHandler);
    try
    {
        if(previousHandler.get() == this)
        {
            return newHandler;
        }
        else if(previousHandler->getConnection() == _connection)
        {
            //
            // If both request handlers point to the same connection, we also
            // update the request handler. See bug ICE-5489 for reasons why
            // this can be useful.
            //
            return newHandler;
        }
    }
    catch(const Ice::Exception&)
    {
        // Ignore.
    }
    return ICE_SHARED_FROM_THIS;
}

AsyncStatus
ConnectionRequestHandler::sendAsyncRequest(const ProxyOutgoingAsyncBasePtr& out)
{
    return out->invokeRemote(_connection, _compress, _response);
}

void
ConnectionRequestHandler::asyncRequestCanceled(const OutgoingAsyncBasePtr& outAsync, const Ice::LocalException& ex)
{
    _connection->asyncRequestCanceled(outAsync, ex);
}

Ice::ConnectionIPtr
ConnectionRequestHandler::getConnection()
{
    return _connection;
}

Ice::ConnectionIPtr
ConnectionRequestHandler::waitForConnection()
{
    return _connection;
}

File: java/src/Ice/src/main/java/com/zeroc/IceInternal/ConnectionRequestHandler.java
//
// Copyright (c) ZeroC, Inc. All rights reserved.
//

package com.zeroc.IceInternal;

public final class ConnectionRequestHandler implements RequestHandler {
    @Override
    public int sendAsyncRequest(ProxyOutgoingAsyncBase out) throws RetryException {
        return out.invokeRemote(_connection, _compress, _response);
    }

    @Override
    public void asyncRequestCanceled(
            OutgoingAsyncBase outgoingAsync, com.zeroc.Ice.LocalException ex) {
        _connection.asyncRequestCanceled(outgoingAsync, ex);
    }

    @Override
    public com.zeroc.Ice.ConnectionI getConnection() {
        return _connection;
    }

    public ConnectionRequestHandler(
            Reference reference, com.zeroc.Ice.ConnectionI connection, boolean compress) {
        _reference = reference;
        _response = _reference.isTwoway();
        _connection = connection;
        _compress = compress;
    }

    private final Reference _reference;
    private final boolean _response;
    private final com.zeroc.Ice.ConnectionI _connection;
    private final boolean _compress;
}
